src/sentry/spans/consumers/process/flusher.py

```diff
@@ -122,7 +123,16 @@ def produce(payload: KafkaPayload) -> None:
                 now = int(time.time()) + current_drift.value
                 flushed_segments = buffer.flush_segments(now=now, max_segments=max_flush_segments)
 
-                if len(flushed_segments) >= max_flush_segments * len(buffer.assigned_shards):
+                segments_per_shard = defaultdict(int)
+                for flushed_segment in flushed_segments.values():
+                    segments_per_shard[flushed_segment.queue_key] += 1
```

Comment:
As an idea, what if we move this logic into the buffer and set a flag on the buffer instance inside of `flush_segments`? Then we can check this flag via some getter like `should_pause()` or `is_backlogged()`.